malcontent-control/application.c

```diff
@@ -215,6 +216,92 @@ mct_application_activate (GApplication *application)
   gtk_window_present (window);
 }
 
+static void
+mct_application_shutdown (GApplication *application)
+{
+  MctApplication *self = MCT_APPLICATION (application);
+  EmtrEventRecorder *recorder;
+  g_autoptr(MctManager) mct_manager = NULL;
+  g_autoptr(GSList) users = NULL;  /* (element-type ActUser) */
+  g_auto(GVariantBuilder) builder = G_VARIANT_BUILDER_INIT (G_VARIANT_TYPE ("aa{sv}"));
+  guint n_entries = 0;
+
+  /* Endless-specific code to send metrics containing all the parental controls
+   * for all users, so we can see how different parental controls features are
+   * being used.
+   *
+   * Serialise the app filter for each user. It’s OK to use blocking calls here,
+   * as the UI is no longer shown.
+   *
+   * See https://phabricator.endlessm.com/T28741#810046 */
+#define MCT_PARENTAL_CONTROLS_EVENT "449ec188-cb7b-45d3-a0ed-291d943b9aa6"
+
+  g_debug ("Gathering parental controls statistics");
+
+  users = act_user_manager_list_users (self->user_manager);
+  mct_manager = mct_manager_new (self->dbus_connection);
+
+  for (GSList *l = users; l != NULL; l = l->next)
+    {
+      ActUser *user = ACT_USER (l->data);
+      g_autoptr(MctAppFilter) filter = NULL;
+      g_autoptr(GError) local_error = NULL;
+      g_autoptr(GVariant) serialised_filter = NULL;
+      gboolean is_administrator;
+      g_auto(GVariantDict) dict = G_VARIANT_DICT_INIT (NULL);
+
+      /* Skip system accounts. */
+      if (act_user_is_system_account (user))
+        continue;
+
+      /* Get the user’s filter. */
+      filter = mct_manager_get_app_filter (mct_manager,
+                                           act_user_get_uid (user),
+                                           MCT_MANAGER_GET_VALUE_FLAGS_NONE,
+                                           NULL,
+                                           &local_error);
+      if (g_error_matches (local_error, MCT_MANAGER_ERROR, MCT_MANAGER_ERROR_DISABLED))
+        {
+          g_debug ("Skipping metrics submission as parental controls are globally disabled");
+          break;
+        }
+      else if (g_error_matches (local_error, MCT_MANAGER_ERROR, MCT_MANAGER_ERROR_PERMISSION_DENIED))
```

Comment:
If we were to send this metric when the app is unlocked, we wouldn't have to worry about this error. As written, we're definitely going to have incomplete data when the app is run by a non-admin (only their own info), but what happens when it's run by an admin? Won't it be incomplete there too if it's been more than 5 minutes (which I think is the timeout for the polkit authorization we get when the app is unlocked)?

Reply:
I agree that the current situation isn’t great, although I think that most users will close the app after making their changes, before the polkit timeout — so that should work fine.

Sending the metrics when the user unlocks the app means we’ll be sending the old values (before they edit anything) not the new ones, which is a bit against the point of this metric.

Reply:
I didn't mean we should send the metric at the moment the app is unlocked. I meant send it while the app is unlocked, though I see how you'd need a heuristic for estimating when the user finished editing the controls (or just send on every edit).

Anyway, you're right that users will probably close the app in short order. So I guess we can leave it as is. And I see there's an AC on the ticket for metrics data being received so if this doesn't work it should be caught.